lib/index.js

```diff
@@ -76,57 +77,269 @@ class AuthPassportLocal {
+  exists (request, kuid) {
+    return this.getCredentialsFromUserId(kuid)
+      .then(credentials => credentials !== null);
+  }
+
+  /**
+   * @param {KuzzleRequest} request
+   * @param {string} username
+   * @param {string} password
+   * @returns {Promise<string|{message: string}>}
+   */
+  verify (request, username, password) {
+    return this.getUsersRepository().get(username)
+      .then(result => {
+        if (result === null) {
+          return Promise.resolve({message: 'wrong username or password'});
+        }
+
+        if (!this.passwordManager.checkPassword(password, result.userPassword)) {
+          return Promise.resolve({message: 'wrong username or password'});
+        }
+
+        return Promise.resolve(result.kuid);
+      });
+  }
+
+  /**
+   * @param {KuzzleRequest} request
+   * @param {object} credentials
+   * @param {string} kuid
+   * @returns {Promise<object>}
+   */
+  create (request, credentials, kuid) {
+    if (!credentials.password) {
+      return Promise.reject(new this.context.errors.BadRequestError('Password needed.'));
     }
+    return this.exists(request, kuid)
+      .then(exists => {
+        if (exists) {
+          return Promise.reject(new this.context.errors.BadRequestError('A strategy already exists for this user.'));
+        }
+
+        return this.getUsersRepository().create({
+          kuid,
+          _id: credentials.username,
+          userPassword: this.passwordManager.encryptPassword(credentials.password)
+        }, {refresh: 'wait_for'})
+          .then(createdDocument => Promise.resolve(this.outputDocument(createdDocument)));
+      });
+  }
 
-    if (!isString(request.input.body.password)) {
-      return callback(new this.context.errors.BadRequestError('Missing or invalid given property: password'));
+  /**
+   * @param {KuzzleRequest} request
+   * @param {object} credentials
+   * @param {string} kuid
+   * @returns {Promise<object>}
+   */
+  update (request, credentials, kuid) {
+    if (credentials.kuid) {
+      return Promise.reject(new this.context.errors.BadRequestError('The request must contain a kuid attribute.'));
     }
+    return this.getCredentialsFromUserId(kuid)
+      .then(document => {
+        const credentialUpdate = {};
+
+        if (document === null) {
+          return Promise.reject(new this.context.errors.BadRequestError('A strategy does not exist for this user.'));
+        }
 
-    if (request.input.body.password.trim() === '') {
-      return callback(new this.context.errors.BadRequestError('Empty password is not allowed'));
+        if (credentials.password) {
+          credentialUpdate.userPassword = this.passwordManager.encryptPassword(credentials.password);
+        }
+
+        // To change the username, we have to create a new (complete) document and remove the old one
+        if (credentials.username && document._id !== credentials.username) {
+          let createdDocument;
+          const oldDocumentId = document._id;
```

Comment:
I mean : we don't need to assign a variable `oldDocumentId`.

and we can replace line 242 below (`return this.getUsersRepository().delete(oldDocumentId);`) by:
```javascript
return this.getUsersRepository().delete(document._id);
```